core/src/main/java/net/adoptopenjdk/icedteaweb/resources/downloader/PrioritizedParallelExecutor.java

```diff
@@ -0,0 +1,220 @@
+        result.setLowerPriority(lowerPriority);
+
+        return result;
+    }
+
+    /**
+     * Wrapper around {@link Callable} which will cancel any lower priority after completing the call successfully.
+     */
+    private static class CancelingCallable<V> implements Callable<V> {
+
+        private final AtomicBoolean complete = new AtomicBoolean(false);
+        private final AtomicReference<Future<V>> lowerPriority = new AtomicReference<>();
+
+        private final Callable<V> delegate;
+
+        private CancelingCallable(Callable<V> delegate) {
+            this.delegate = delegate;
+        }
+
+        /**
+         * Set a lower priority to be canceled.
+         * This is not passed in the constructor as the value is not known then.
+         * <p>
+         * If this callable has already been completed successfully then the passed future will be canceled immediately.
+         *
+         * @param lowerPriority the lower priority future to be canceled after completing this callable successfully.
+         */
+        void setLowerPriority(Future<V> lowerPriority) {
+            this.lowerPriority.set(lowerPriority);
+            if (complete.get()) {
+                cancelLowerPriority();
+            }
+        }
+
+        @Override
+        public V call() throws Exception {
+            final V result = delegate.call();
+            complete.set(true);
+            cancelLowerPriority();
+            return result;
+        }
+
+        private void cancelLowerPriority() {
+            final Future<V> lowerPriority = this.lowerPriority.get();
+            if (lowerPriority != null) {
+                lowerPriority.cancel(true);
+            }
+        }
+    }
+
+    /**
+     * Wrapper around {@link Future} which allows to construct a chain of futures.
+     * Lower priority futures are canceled if this future is canceled.
+     * If this future fails the result of the lower priority is returned.
+     */
+    private static class ChainableFuture<V> implements Future<V> {
+
+        private final AtomicBoolean terminatedExceptionally = new AtomicBoolean(false);
+        private final CompletableFuture<V> lowerPriorityResult = new CompletableFuture<>();
+        private final AtomicReference<Future<V>> lowerPriority = new AtomicReference<>();
+
+        private final Future<V> delegate;
+
+        private ChainableFuture(Future<V> delegate) {
+            this.delegate = delegate;
+        }
+
+        /**
+         * Set a lower priority to be canceled.
+         * This is not passed in the constructor as the value is not known then.
+         *
+         * @param lowerPriority the lower priority future to be canceled this future is canceled.
+         */
+        void setLowerPriority(Future<V> lowerPriority) {
+            this.lowerPriority.set(lowerPriority);
+
+            if (terminatedExceptionally.get()) {
+                Executors.newSingleThreadExecutor().execute(() -> {
```

Comment:
mmmh, I do not know if `Executors.newSingleThreadExecutor()` is a good idea. The basic idea behind this PR was to have a list that is executed by a concrete executor.

Reply:
how else would you prevent the call to `setLowerPriority` from being blocking

Reply:
I do not say that I have the perfect solution ;) I assume that at least you should pass the executor up to this place. Do not know how much complexity this will add.